myia/parser.py

```diff
@@ -311,13 +381,73 @@ def process_Return(self, block: 'Block', node: ast.Return) -> 'Block':
     def process_Assign(self, block: 'Block', node: ast.Assign) -> 'Block':
         """Process an assignment."""
         anf_node = self.process_node(block, node.value)
-        if len(node.targets) == 1 and isinstance(node.targets[0], ast.Name):
-            anf_node.debug.name = node.targets[0].id
-            block.write(node.targets[0].id, anf_node)
-        else:
-            raise NotImplementedError(node.targets)
+        targ, = node.targets
+
+        def write(targ, anf_node):
+
+            if isinstance(targ, ast.Name):
+                # CASE: x = value
+                anf_node.debug.name = targ.id
+                block.write(targ.id, anf_node)
+
+            elif isinstance(targ, ast.Tuple):
+                # CASE: x, y = value
+                for i, elt in enumerate(targ.elts):
+                    op = self.environment.map(operator.getitem)
+                    new_node = Apply([op, anf_node, Constant(i)], block.graph)
+                    write(elt, new_node)
+
+            elif isinstance(targ, ast.Subscript):
+                if isinstance(targ.value, ast.Name):
+                    # CASE: x[y] = value
+                    op = self.environment.map(operator.setitem)
+                    obj = self.process_node(block, targ.value)
+                    idx = self.process_node(block, targ.slice)
+                    new_node = Apply([op, obj, idx, anf_node], block.graph)
+                    write(targ.value, new_node)
+                else:
+                    # UNSUPPORTED: f()[x] = value
+                    raise NotImplementedError(
+                        "You can only set a slice on a variable."
+                    )  # pragma: no cover
```

Comment:
I don't think we should support subscript assignment at all since it changes the semantic from an inplace modification to a copy-and-modify.

In the face on aliasing, it will break code.  We attempt to discover a simple aliasing case when the value is not a local variable, but we cannot be certain that it is safe with only the information that the parser has here.  This would need flow analysis to make sure it's ok.

At that point it would be majorly confusing for the user to have this syntax work sometimes and not others for seemingly arbitrary reasons.

Reply:
We've discussed this in the past. I think there are pros and cons for each approach, let me rehash them:

* `x[i] = v` is much more user-friendly than `x = tensor.set_subtensor, x, i , v)`
* It will make it a lot easier to compile existing functions with minimal changes; otherwise I need to go through all of my NumPy code changing subscript assignments to their functional equivalent. This is problematic if one of our goals is for `@myia` being all that is needed to get speedups.
* This really only matters if `x` is used afterwards in a different scope
* If we want to support e.g. `print` statements or `rand` then we will need functions with side-effects. Whatever solution we use in those cases might be used here as well.
* We might be able to do some clever things and rewrite this:
```python
def f(x):
    x[i] = v
    return sum(x)

s = f(x)
print(x)
```
do this
```python
def f(x):
    x[i] = v
    return x, sum(x)

x, s = f(x)
print(x)
```

On the other hand:

* The rewriting approach I mentioned above might get pretty tricky in the presence of HOFs and closures.
* If we don't rewrite, then diverging from Python behavior is risky, and no matter how clearly we put it in the documentation that `x[i] = v` should be read as `x is x with x[i] = v` some people will be surprised by this behavior.
* Using a function call is more explicit, less magic, which is good

I'm a bit on the fence. If we can come up with a "functional update" syntax that isn't too bad (like OCaml's), then I could be more easily swayed to the explicit approach. Maybe something like `x = x.set(i, v)`?

Reply:
I think it would be premature to try to settle this issue now, so what I'm going to do is pull out the support here and leave it in a new branch so that we can apply it later if we decide that this is something we want.

Alternatively, we could leave it and have a flag in the syntax checker to turn it on or off, but that's not in clean yet.

Reply:
Just my two cents here:
- I think supporting `x += 1` should not be any trouble, since it can legitimately be understood by Python as a reassignment, `x = x + 1`
- No matter which syntax we come up with, numerous users will expect `x.set(i, v)` or `set_subtensor(x[i], v)` to perform the assignment in itself, and expect `x` to hold the new value subsequently (even if they do not assign the return value, or assign it to `new_x`).
- On the other hand, users would also expect that to actually produce side effects. For instance, they would expect:
```python
y = x
x[i] = v
assert y[i] == v
```
So unless we can correctly detect that case and forbid pseudo-inplace updates when the original value is aliased, maybe we're better off forbidding it altogether.

Reply:
`x += y` amounts to `x = x.__iadd__(y)`, which is a destructive operation for several types, so we'd still have to forbid it for lists, ndarrays, and so on.

Reply:
My position is not that we should not try to support the `x[i] = v` syntax.  It's rather that we should make sure to preserve the semantics if we do.  Until we can assure that, we should not claim to support it.

In the long term I think it would be possible to do a transform like @bartvm showed to propagate the changes and at that point we should introduce the support.